### PR TITLE
(Resubmission) Improve Flatten to avoid using dynamic shapes when possible

### PR DESCRIPTION
**Pull request by trevor-m**
Fixes issues @ebrevdo pointed out with #30380 
I also added a unit test for the int32 overflow situation.
See original PR for explanation of overall changes.

### CODE REVIEW

File: tensorflow/python/keras/layers/core.py [hunk 1]
```diff
@@ -580,9 +581,21 @@ def call(self, inputs):
...
+    if input_shape[1:].is_fully_defined():
+      flattened_dim = tensor_shape.dimension_value(
+          np.prod(input_shape[1:], dtype=int))
+      # Temporary fix for integer overflow issue.
+      if flattened_dim > np.iinfo(np.int32).max:
+        shape_dtype = dtypes.int64
+      else:
+        shape_dtype = dtypes.int32
+      outputs = array_ops.reshape(
+          inputs, constant_op.constant((-1, flattened_dim), shape_dtype))
```
Comment (ebrevdo) [inline tensorflow/python/keras/layers/core.py]: constant(..., dtype=shape_dtype)

File: tensorflow/python/keras/layers/core.py [hunk 2]
```diff
@@ -580,9 +581,21 @@ def call(self, inputs):
...
+    else:
+      outputs = array_ops.reshape(
+          inputs, (tensor_shape.dimension_value(inputs.shape[0]) or
```
Comment (ebrevdo) [inline tensorflow/python/keras/layers/core.py]: might as well add the check here for the size of `dimension_value(inputs.shape[0])` as well.  i expect a similar issue could come up with e.g. a batch size that's overly large.

**Comment by ebrevdo** [comment]
Thanks for fixing!

**Comment by trevor-m** [comment]
Thanks for the improvements @ebrevdo . Let me know if you have any ideas to shorten the code since it's gotten a bit long now.

File: tensorflow/python/keras/layers/core.py [hunk 3]
```diff
@@ -580,9 +581,31 @@ def call(self, inputs):
...
+      batch_size = tensor_shape.dimension_value(inputs.shape[0])
+      if batch_size:
+        # Temporary fix for integer overflow issue.
+        if batch_size > np.iinfo(np.int32).max:
+          shape_dtype = dtypes.int64
+        else:
+          shape_dtype = dtypes.int32
+        outputs = array_ops.reshape(
+            inputs, constant_op.constant(
+                (tensor_shape.dimension_value(inputs.shape[0]), -1),
```
Comment (ebrevdo) [inline tensorflow/python/keras/layers/core.py]: should this be constant_op.constant([batch_size, -1], dtype=...)

?

Comment (trevor-m) [reply tensorflow/python/keras/layers/core.py]: Oops, thanks for catching that

**Review by ebrevdo — APPROVED** [review]
thanks!

**Comment by trevor-m** [comment]
Hey @ebrevdo it looks like the PR was reverted again?

**Comment by ebrevdo** [comment]
yes :(  for some reason the shape inference code on the *python* side is
truncating in *windows* (this affects the unit test).  i'll follow up with
our internal team on how we can replicate this in windows.  if you have a
windows box, let us know as you may be able to do some debugging by just
running the unit test with your PR in.

**Comment by ebrevdo** [comment]
My suggestion for now is to disable the unit test in windows; easiest way to check:

import platform
is_windows = any(platform.win32_ver())

and self.skipTest if is_windows.

**Comment by ebrevdo** [comment]
@mrry know who might have time to figure out why shape inference code (either the c++ or swig) is truncating int32 values in windows?  i tried debugging it but ran into too many issues with the tooling.

**Comment by ebrevdo** [comment]
@trevor-m sorry for the delay.  i'm hoping that final change will allow this PR to stick :)

**Comment by pkanwar23** [comment]
ebrevdo@- I'm not clear on what you are suggesting. Should we disable the unit test on windows? Or are you saying you've uncovered a different path to make the PR stick?

**Comment by ebrevdo** [comment]
disabling the unit test on windows is the best choice right now.
